### PR TITLE
Fix up the set of Windows compatibility libraries exposed in WindowsDesktop

### PR DESCRIPTION
**Pull request by dagood**
For https://github.com/dotnet/core-setup/issues/7290.

This PR assumes that we should use `System.IO.Pipes.AccessControl` `4.5.1`. (https://github.com/dotnet/core-setup/issues/7290#issuecomment-513926657)

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -88,6 +88,10 @@
     <MicrosoftDiaSymReaderNativePackageVersion>1.7.0</MicrosoftDiaSymReaderNativePackageVersion>
     <!-- Infrastructure and test-only. -->
     <MicrosoftSourceLinkVersion>1.0.0-beta2-18618-05</MicrosoftSourceLinkVersion>
+    <!-- TODO: Sort these into the above. These are out of the way at the moment to make it easier to resolve conflicts. -->
```
Comment (dagood) [inline eng/Versions.props]: I'll do these with my merge commit. (If this PR is the way to go.)

File: src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj [hunk 2]
```diff
@@ -5,28 +5,31 @@
...
+    <PackageReference Include="System.Threading.AccessControl" Version="$(SystemThreadingAccessControlVersion)" />
     <PackageReference Include="System.Windows.Extensions" Version="$(SystemWindowsExtensionsPackageVersion)" />
   </ItemGroup>
 
   <ItemGroup>
-    <RuntimeOnlyPackageReference Include="System.Diagnostics.EventLog" Version="$(SystemDiagnosticsEventLogVersion)" />
-    <RuntimeOnlyPackageReference Include="System.DirectoryServices" Version="$(SystemDirectoryServicesVersion)" />
-    <RuntimeOnlyPackageReference Include="System.IO.FileSystem.AccessControl" Version="$(SystemIOFileSystemAccessControlVersion)" />
-    <RuntimeOnlyPackageReference Include="System.Threading.AccessControl" Version="$(SystemThreadingAccessControlVersion)" />
+    <RuntimeOnlyPackageReference Include="System.Security.Permissions" Version="$(SystemSecurityPermissionsVersion)" />
```
Comment (dagood) [inline src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: I removed this entirely with my first commit, but that isn't right: this package is depended on by a few of the others, so the ref assembly was pulled in anyway. This caused CI failure when the ref didn't have `Profile` classification. I have to include this package reference to mark the exclusion.

File: src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj [hunk 3]
```diff
@@ -51,6 +51,9 @@
 
     <!-- We don't have a ref assembly for DirectWriteForwarder -->
     <IgnoredReference Condition="'$(PackageTargetRuntime)' == ''" Include="DirectWriteForwarder" />
+
+    <!-- We want to intentionally exclude the ref assembly for System.Security.Permissions. -->
+    <IgnoredReference Condition="'$(PackageTargetRuntime)' == ''" Include="System.Security.Permissions" />
```
Comment (ericstj) [inline src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: Careful here: who's referencing this?  I would have expected that reference to be gone.  We should wait to remove it until it is.

File: src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj [hunk 4]
```diff
@@ -5,28 +5,34 @@
...
+    <PackageReference Include="System.Threading.AccessControl" Version="$(SystemThreadingAccessControlVersion)" />
     <PackageReference Include="System.Windows.Extensions" Version="$(SystemWindowsExtensionsPackageVersion)" />
   </ItemGroup>
 
   <ItemGroup>
-    <RuntimeOnlyPackageReference Include="System.Diagnostics.EventLog" Version="$(SystemDiagnosticsEventLogVersion)" />
-    <RuntimeOnlyPackageReference Include="System.DirectoryServices" Version="$(SystemDirectoryServicesVersion)" />
-    <RuntimeOnlyPackageReference Include="System.IO.FileSystem.AccessControl" Version="$(SystemIOFileSystemAccessControlVersion)" />
-    <RuntimeOnlyPackageReference Include="System.Threading.AccessControl" Version="$(SystemThreadingAccessControlVersion)" />
+    <RefOnlyPackageReference Include="System.IO.Pipes.AccessControl" Version="$(SystemIOPipesAccessControl)" />
```
Comment (ericstj) [inline src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: Why did you need to explicitly make it RefOnly?  Can we instead make conflict resolution do the work to remove anything that would conflict with the base shared-framework and lose?

File: src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj (hunk 3 above)
Comment (dagood) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: I see... when I remove the ignore, the list is:

```
Assembly 'System.Xaml' is missing dependency 'System.Security.Permissions'
Assembly 'System.Configuration.ConfigurationManager' is missing dependency 'System.Security.Permissions'
Assembly 'System.Security.Cryptography.Xml' is missing dependency 'System.Security.Permissions'
Assembly 'WindowsBase' is missing dependency 'System.Security.Permissions'
Assembly 'System.Windows.Forms.Design.Editors' is missing dependency 'System.Security.Permissions'
Assembly 'System.Drawing' is missing dependency 'System.Security.Permissions'
Assembly 'System.DirectoryServices' is missing dependency 'System.Security.Permissions'
```

File: src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj (hunk 4 above)
Comment (dagood) [reply src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: AFAIK we don't have tooling like that yet, I created https://github.com/dotnet/core-setup/issues/7340 to track that. Definitely seems possible.

Comment (ericstj) [reply src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: What do you mean by "tooling like that"  It should be built into the SDK.  You just need to set a property that points to the platform manifest you generated in this repository.

Comment (ericstj) [reply src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: Scratch that, an item.  Try just setting PackageConflictPlatformManifests item.

File: src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj (hunk 3 above)
Comment (ericstj) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: We need to determine why this stuff is pulling it in.  Dangling type-forwards are OK, actual type references are not.

File: src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj (hunk 4 above)
Comment (dagood) [reply src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: That's exactly what I mean, wasn't aware of that. :smile: I'll give it a try.

File: src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj (hunk 3 above)
Comment (dagood) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: We haven't had an update from WPF for a while, and it looks like there are some relevant changes we don't have yet: https://github.com/dotnet/wpf/compare/ea72094ebe0de0da03c591752559739d6fb67f03...master.

Comment (dagood) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: Ah... `dotnet-wpf-int` isn't producing for 5.0 yet. Filed https://github.com/dotnet/core-setup/issues/7343 to figure out how to get these updated. For now, I'll keep `System.Security.Permissions` so the rest can be mergeable.

Comment (ericstj) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: That's not the only cause for this.  Some of these other dependencies are concerning.  I suspect we'll need to suppress this, since even once everything is fixed WindowsBase will have typeforwards to System.Security.Permission, but we need to examine all the references before suppressing to find out if there is other work to do to safely omit this.

Comment (dagood) [reply src/pkg/projects/windowsdesktop/pkg/Microsoft.WindowsDesktop.App.pkgproj]: Filed https://github.com/dotnet/core-setup/issues/7344 to follow up on this, I'll keep this PR to what seems doable before then. Thanks for the explanation, this makes sense. (I don't know what the best way to check the references is to get this done myself.)

File: src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj (hunk 4 above)
Comment (dagood) [reply src/pkg/projects/windowsdesktop/src/windowsdesktop.depproj]: Seems to work as expected! Thanks!

I had to change `GetNETCoreAppIgnoredReference` a bit for closure validation: when verifying the runtime package, it ignores netcoreapp *refs*, not *runtimes*, causing [a number of missing reference errors](https://gist.github.com/dagood/2d9ea2ed775edf884e2ff6f3dfc66bf0). I think that was just a bug that happened to rely on the asset duplication we had.

**Comment by dagood** [comment]
I diffed the data files (runtimelist/frameworklist) to get a summary of the results: https://gist.github.com/dagood/5f328e6148acd7cc8ee1afcc6c60f814. All the removals are in the NETCoreApp sfx, other changes look like what we intended.

**Comment by dagood** [comment]
I didn't rerun crossgen in my local build, which looks like it's failing now. Looking into it.

**Comment by ericstj** [comment]
It's due to WindowsBase removal:
```
  Error generating PDB for 'F:\workspace\_work\1\s\artifacts\bin\win-x64.Debug\crossgen\windowsdesktop\runtimes/win-x64/lib/netcoreapp5.0/PresentationFramework.Aero.dll': Could not load type 'System.Windows.DependencyObject' from assembly 'WindowsBase, Version=4.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'.
  Microsoft (R) CoreCLR Native Image Generator - Version 4.5.30319.0
  Copyright (c) Microsoft Corporation.  All rights reserved.
```

**Comment by dagood** [comment]
I'm going to apply this PR to `release/3.0` directly to make progress. https://github.com/dotnet/wpf/issues/1351 tracks fixing the WindowsBase conflict resolution problem for 5.0. (Some discussion about what's going on there in the gist: https://gist.github.com/dagood/5f328e6148acd7cc8ee1afcc6c60f814.)

**Comment by dleeapho** [comment]
Should probably move this PR to the new dotnet/windowsdesktop repo.

**Comment by dagood** [comment]
This PR makes changes to some files now in the sharedfx tooling SDK and relies on the fact that netcoreapp and windowsdesktop are both in this repo, so I can't actually move the PR over. However, I opened https://github.com/dotnet/windowsdesktop/issues/7 to track this and I'm closing this PR.
